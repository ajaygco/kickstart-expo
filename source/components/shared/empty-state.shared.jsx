// Modules
import React from "react";
import { Text, View } from "react-native";

// Shared
import { If } from "components/shared/if.shared";

// Component: Presentation
export const EmptyStateUi = ({ title, message, illustration }) => {
  const shouldShowIllustration = React.isValidElement(illustration);

  return (
    <View className="w-full items-center justify-center p-16">
      <If condition={shouldShowIllustration}>
        <View className="w-64 h-64 flex items-center justify-center mb-8">
          {illustration}
        </View>
      </If>

      <If condition={title}>
        <Text
          allowFontScaling={false}
          className="text-xl text-gray-800 font-sans font-bold mb-2"
        >
          {title}
        </Text>
      </If>

      <If condition={message}>
        <Text
          allowFontScaling={false}
          className="text-base text-gray-600 text-center font-sans"
        >
          {message}
        </Text>
      </If>
    </View>
  );
};

// Component: Logic
export const EmptyStateLogic = (props) => {
  return <EmptyStateUi {...props} />;
};

// Component: Memo
export const EmptyState = React.memo((props) => {
  return <EmptyStateLogic {...props} />;
});
