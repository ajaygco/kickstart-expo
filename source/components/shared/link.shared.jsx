// Modules
import React from "react";
import { Linking, Pressable } from "react-native";

// Helpers
import { Alert } from "helpers/alert.helper";
import { captureException } from "helpers/analytics.helper";

// Component: Presentation
export const LinkUi = ({ onPress, children, ...props }) => {
  return (
    <Pressable hitSlop={12} onPress={onPress} {...props}>
      {children}
    </Pressable>
  );
};

// Component: Logic
export const LinkLogic = ({ to, ...props }) => {
  // On Press
  const onPress = async () => {
    try {
      await Linking.openURL(to);
    } catch (error) {
      captureException(error);

      Alert.error({
        title: "Error",
        message: "Couldn't open link",
      });
    }
  };

  return <LinkUi onPress={onPress} {...props} />;
};

// Component: Memo
export const Link = React.memo((props) => {
  return <LinkLogic {...props} />;
});
