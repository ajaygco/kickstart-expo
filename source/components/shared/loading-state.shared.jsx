// Modules
import React from "react";
import { ActivityIndicator, View } from "react-native";

// Component: Presentation
export const LoadingStateUi = (props) => {
  return (
    <View className="items-center justify-center">
      <ActivityIndicator {...props} />
    </View>
  );
};

// Component: Logic
export const LoadingStateLogic = (props) => {
  return <LoadingStateUi {...props} />;
};

// Component: Memo
export const LoadingState = React.memo((props) => {
  return <LoadingStateLogic {...props} />;
});
