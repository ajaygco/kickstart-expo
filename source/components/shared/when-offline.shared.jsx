// Modules
import React from "react";
import { useIsConnected } from "react-native-offline";

// Component: Presentation
export const WhenOfflineUi = ({ isConnected, children }) => {
  if (isConnected) {
    return;
  }

  return children;
};

// Component: Logic
export const WhenOffline = ({ ...props }) => {
  // Hooks
  const isDeviceConnected = useIsConnected();

  return <WhenOfflineUi isConnected={isDeviceConnected} {...props} />;
};
