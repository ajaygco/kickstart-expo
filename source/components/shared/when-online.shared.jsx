// Modules
import React from "react";
import { useIsConnected } from "react-native-offline";

// Component: Presentation
export const WhenOnlineUi = ({ isConnected, children }) => {
  if (!isConnected) {
    return;
  }

  return children;
};

// Component: Logic
export const WhenOnline = (props) => {
  const isConnected = useIsConnected();

  return <WhenOnlineUi isConnected={isConnected} {...props} />;
};
