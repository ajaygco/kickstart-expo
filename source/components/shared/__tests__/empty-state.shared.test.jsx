// bunx jest source/components/shared/__tests__/empty-state.shared.test.jsx

import React from "react";
import { render } from "@testing-library/react-native";
import { View } from "react-native";

import { EmptyState } from "components/shared/empty-state.shared";

describe("EmptyState", () => {
  it("should render correctly", () => {
    const wrapper = render(
      <EmptyState
        title="Title"
        message="Some Message"
        illustration={<View testID="actingIllustration" />}
      />,
    );

    const titleElement = wrapper.getByText("Title");
    const MessageElement = wrapper.getByText("Some Message");

    const illustration = wrapper.getByTestId("actingIllustration");

    expect(titleElement).toBeTruthy();
    expect(MessageElement).toBeTruthy();

    expect(illustration).toBeTruthy();

    expect(wrapper).toMatchSnapshot();
  });

  it("should not render illustration when empty", () => {
    const wrapper = render(<EmptyState title="Title" message="Some Message" />);

    const titleElement = wrapper.getByText("Title");
    const MessageElement = wrapper.getByText("Some Message");

    const illustration = wrapper.queryByTestId("actingIllustration");

    expect(titleElement).toBeTruthy();
    expect(MessageElement).toBeTruthy();

    expect(illustration).toBeFalsy();

    expect(wrapper).toMatchSnapshot();
  });
});
