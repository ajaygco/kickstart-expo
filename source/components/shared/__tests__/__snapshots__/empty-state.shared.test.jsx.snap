// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmptyState should not render illustration when empty 1`] = `
<View
  style={
    [
      {
        "width": "100%",
      },
      {
        "alignItems": "center",
      },
      {
        "justifyContent": "center",
      },
      {
        "paddingBottom": 64,
        "paddingLeft": 64,
        "paddingRight": 64,
        "paddingTop": 64,
      },
    ]
  }
>
  <Text
    allowFontScaling={false}
    style={
      [
        {
          "fontSize": 20,
          "lineHeight": 28,
        },
        {
          "color": "#1f2937",
        },
        {
          "fontFamily": "Figtree",
        },
        {
          "fontWeight": "700",
        },
        {
          "marginBottom": 8,
        },
      ]
    }
  >
    Title
  </Text>
  <Text
    allowFontScaling={false}
    style={
      [
        {
          "fontSize": 16,
          "lineHeight": 24,
        },
        {
          "color": "#4b5563",
        },
        {
          "textAlign": "center",
        },
        {
          "fontFamily": "Figtree",
        },
      ]
    }
  >
    Some Message
  </Text>
</View>
`;

exports[`EmptyState should render correctly 1`] = `
<View
  style={
    [
      {
        "width": "100%",
      },
      {
        "alignItems": "center",
      },
      {
        "justifyContent": "center",
      },
      {
        "paddingBottom": 64,
        "paddingLeft": 64,
        "paddingRight": 64,
        "paddingTop": 64,
      },
    ]
  }
>
  <View
    style={
      [
        {
          "width": 256,
        },
        {
          "height": 256,
        },
        {
          "display": "flex",
        },
        {
          "alignItems": "center",
        },
        {
          "justifyContent": "center",
        },
        {
          "marginBottom": 32,
        },
      ]
    }
  >
    <View
      testID="actingIllustration"
    />
  </View>
  <Text
    allowFontScaling={false}
    style={
      [
        {
          "fontSize": 20,
          "lineHeight": 28,
        },
        {
          "color": "#1f2937",
        },
        {
          "fontFamily": "Figtree",
        },
        {
          "fontWeight": "700",
        },
        {
          "marginBottom": 8,
        },
      ]
    }
  >
    Title
  </Text>
  <Text
    allowFontScaling={false}
    style={
      [
        {
          "fontSize": 16,
          "lineHeight": 24,
        },
        {
          "color": "#4b5563",
        },
        {
          "textAlign": "center",
        },
        {
          "fontFamily": "Figtree",
        },
      ]
    }
  >
    Some Message
  </Text>
</View>
`;
